### ISSUE TITLE
Preventing race conditions with CRON jobs

### ISSUE THREAD

**Issue by Cpaczek** [strapi/strapi#15634]
As requested by @petersg83 https://github.com/strapi/strapi/pull/15590 i've opened a new issue to discuss the best way to prevent race conditions with CRON jobs for horizontally scaled Strapi instance. 

Currently, all CRON jobs will fire on all Strapi instances no matter what. The best way to solve this would be to push Cron jobs to a table and lock that table, when a Cron needs to run only 1 instance will get the lock and can run the code. I could probably work on this in a couple of weeks after I hear input about this implementation and nail down the details a bit more.

**Comment by sjoukedv** [comment]
I agree a (table) lock would be a resolution for the race condition.

It would be great if we can add another flag that specifies if the cron job should run once across all instances or only once, e.g. by storing the date of the latest run and a boolean column that stores if the job did run already. 

```
IF !select(cron_date, did_run) || allow_multiple_crons DO 
  // <-- remove previous cron_date, did_run entry from table
  // <-- do task
```

**Comment by Marc-Roig** [comment]
Looking a bit on how to lock tables, seems mysql discourages to do that when using transactions:
> LOCK TABLES is not transaction-safe and implicitly commits any active transaction before attempting to lock the tables.

Also, it seems knex does not support locking tables straight out of the box either. Maybe with some raw quereis it could be possible.

> It would be great if we can add another flag that specifies if the cron job should run once across all instances or only once, e.g. by storing the date of the latest run and a boolean column that stores if the job did run already.
That would be a nice feature indeed, I wonder if databases have some sort of timed alarm that we could use too.

**Comment by derrickmehaffy** [comment]
> Looking a bit on how to lock tables, seems mysql discourages to do that when using transactions:
> > LOCK TABLES is not transaction-safe and implicitly commits any active transaction before attempting to lock the tables.
> 
> Still, there might be plenty of other mechanisms we could do. I agree this could be a good way.
> 
> > It would be great if we can add another flag that specifies if the cron job should run once across all instances or only once, e.g. by storing the date of the latest run and a boolean column that stores if the job did run already.
> 
> 
> That would be a nice feature indeed, I wonder if databases have some sort of timed alarm that we could use too. 
> 

Most do yes, albeit it I'm not sure we would add it natively but this is specifically what software like Redis was designed for: https://redis.com/redis-best-practices/communication-patterns/redlock/

As for the DB crons:
https://github.com/citusdata/pg_cron
https://dev.mysql.com/doc/refman/8.0/en/events-overview.html
https://mariadb.com/kb/en/event-scheduler/

**Comment by Marc-Roig** [comment]
This could be a provider-cron system where you select in the config the interval time and which controller to hit

**Comment by derrickmehaffy** [comment]
https://github.com/strapi-community/strapi-plugin-redis

Heads up btw, the code is shit because I'm it's maintainer 😂

**Comment by Cpaczek** [comment]
Maybe cron should be a plugin? Then we could have a few different implementations instead of cramming multiple solutions into Strapi core? i.e you could have a plugin that uses redis, one that uses table locking for the dbs that support it and then just keep the current one in strapi. A plugin could easily override the corn service functions during the register phase allowing for drop in plug and play.

**Comment by sjoukedv** [comment]
> > Looking a bit on how to lock tables, seems mysql discourages to do that when using transactions:
> > > LOCK TABLES is not transaction-safe and implicitly commits any active transaction before attempting to lock the tables.
> > 
> > 
> > Still, there might be plenty of other mechanisms we could do. I agree this could be a good way.
> > > It would be great if we can add another flag that specifies if the cron job should run once across all instances or only once, e.g. by storing the date of the latest run and a boolean column that stores if the job did run already.
> > 
> > 
> > That would be a nice feature indeed, I wonder if databases have some sort of timed alarm that we could use too.
> 
> Most do yes, albeit it I'm not sure we would add it natively but this is specifically what software like Redis was designed for: https://redis.com/redis-best-practices/communication-patterns/redlock/
> 
> As for the DB crons: https://github.com/citusdata/pg_cron https://dev.mysql.com/doc/refman/8.0/en/events-overview.html https://mariadb.com/kb/en/event-scheduler/

What is important for my use case that it (1) does not need an extra external service like redis besides the 'main' database, and (2) is not dependent on database extensions/modules. My provider supports [pg_cron](https://docs.digitalocean.com/products/databases/postgresql/details/supported-extensions/), but we can't assume this for all managed databases out there. In my opinion there is nothing wrong with using the current `node-schedule` on each instance as long as there is a mechanism to avoid race conditions and control if it runs once or on every instance (`n` executions would be nice).

**Comment by sjoukedv** [comment]
> 

I am assuming the supported databases taken from [here](https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/installation/cli.html#preparing-the-installation). If we can lock a table perhaps its sufficient for an implementation (this is different from using a native database scheduling feature).

PSQL: https://www.postgresql.org/docs/current/explicit-locking.html
MySQL: https://dev.mysql.com/doc/refman/8.0/en/lock-tables.html
SQLite: https://www.sqlite.org/lockingv3.html
MariaDB: https://mariadb.com/kb/en/lock-tables/

Didn't go over every doc in detail, but I can imagine there can be a different interpretation of guaranteed... The core shouldn't get bloated with functionality that majority doesn't use, writing a separate plugin for it avoids that.

**Comment by Cpaczek** [comment]
One other issue with table locks is that they require a table, which means for each Cron job we would have to make a new table in order to get the lock. Creating tables dynamically like this isn't really a practice that strapi supports.

**Comment by derrickmehaffy** [comment]
> One other issue with table locks is that they require a table, which means for each Cron job we would have to make a new table in order to get the lock. Creating tables dynamically like this isn't really a practice that strapi supports. 

Exactly this and for speed Redis is wayyyyyyy faster than any SQL database for something like this. Redlock (what was implemented in that plugin) is specifically designed for use-cases like this.

https://redis.com/redis-best-practices/communication-patterns/redlock/

**Comment by himanshugogroup** [comment]
One approach is if all cron job from different instance of server triggered at same time create a lock from all jobs but before that add random delay which suits your use case doing this one of service will always have least delay. Doing so one job will get job done which is required for case like send report email at 5am or create resources for next day in advance.

Tip:- For avoid delay collision store that delay in redis also. So no other job from other instance can use it.
